src/vmm/src/persist.rs

```diff
@@ -522,16 +497,77 @@ pub enum GuestMemoryFromUffdError {
     Connect(#[from] std::io::Error),
     /// Failed to sends file descriptor: {0}
     Send(#[from] vmm_sys_util::errno::Error),
+    /// Cannot restore hugetlbfs backed snapshot when using Secret Freedom.
+    HugetlbfsSnapshot,
+}
+
+// TODO remove these when the UFFD crate supports minor faults for guest_memfd
+const UFFDIO_REGISTER_MODE_MINOR: u64 = 1 << 2;
+const UFFDIO_REGISTER: u64 = 0xc020aa00;
+
+use std::ffi::c_void;
+
+use libc::ioctl;
+#[repr(C)]
+struct UffdRegister {
+    range: UffdRange,
+    mode: u64,
+    ioctls: u64,
+}
+
+#[repr(C)]
+struct UffdRange {
+    start: u64,
+    len: u64,
+}
+
+fn guest_memory_uffd_register_minor(guest_memory: &[GuestRegionMmap], uffd: &Uffd) {
+    for mem_region in guest_memory.iter() {
+        let uffd_register = UffdRegister {
+            range: UffdRange {
+                start: mem_region.as_ptr() as u64,
+                len: mem_region.size() as u64,
+            },
+            mode: UFFDIO_REGISTER_MODE_MINOR,
+            ioctls: 0,
+        };
+
+        // SAFETY: ioctl is safe. Called with a valid UFFD object and a valid uffd_register
+        // structure.
+        #[allow(clippy::cast_possible_truncation)]
+        #[allow(clippy::ptr_as_ptr)]
+        let result = unsafe {
+            ioctl(
+                uffd.as_raw_fd(),
+                UFFDIO_REGISTER as _,
+                &uffd_register as *const _ as *const c_void,
+            )
+        };
+
+        if result == -1 {
+            panic!(
+                "Failed to register UFFD minor: {}",
+                io::Error::last_os_error()
+            );
+        }
+    }
```

Comment:
Based on what we talked about the other day how uffd-rs does not need to support UFFDIO_REGISTER_MODE_MINOR for us to be able to use it with their `register()` API:

```suggestion

use vm_memory::VolatileMemory;

fn guest_memory_uffd_register_minor(guest_memory: &[GuestRegionMmap], uffd: &Uffd) {
    for mem_region in guest_memory.iter() {
        uffd.register_with_mode(mem_region.as_ptr().cast(), mem_region.len(), RegisterMode::from_bits_retain(UFFDIO_REGISTER_MODE_MINOR))
            .unwrap();
    }
}
```